### PR TITLE
Args parser does not use memorize

### PR DESCRIPTION
**Pull request by scharf**
`ves` does not work with the master version of `moxb`, because it complains about `lodash.memoize`. This problem does probably not happen, if we would install `moxb` the normal way. But the combination of `lerna` and linked packages causes problems.

However, I was looking why `moxb` needs `lodash.memoize`: it is needed by `UrlArgTypes.ts`.

Looking at the tests I discovered some inconsistencies: 

- sometimes the test requires the result of `parse()` to be the same, and sometimes it is not the same
- for unordered arrays, the order should not matter when `parse()` is called (the result should always be the same)
- the `defaultValue` was ignored for attray types

So, I thought, I would fix it (see PR #47), but that turns out to be more complicated. Therefore, I decided to drop the constraint (and `lodash.memoize`), that the args are the same and reduced it to just be equal.

I think there is a danger in `lodash.memoize` if we are not in a purely functional environment: if you change the returned array, then you change the cached value and all kind of evil things could happen:

```ts
const args = parse('a,b', []);
args[1] = 'OOPS';
parse('a,b', []); ==> ['a', 'OOPS']!
```

see [broken test in #47](https://github.com/moxb/moxb/pull/47/commits/5cadbdbec04bd886095abd31ad94697ae386ae6)

### CODE REVIEW

File: packages/moxb/src/routing/url-arg/UrlArgTypes.ts [hunk 1]
```diff
@@ -14,49 +13,33 @@ export const URLARG_TYPE_STRING: UrlArgTypeDef<string> = {
...
-        v1
-            .slice()
-            .sort()
-            .join(',') ===
-        v2
-            .slice()
-            .sort()
-            .join(','),
-    format: v => v.join(','),
+    getParser: () => (v, defaultValue) => (v.length ? v.split(',') : defaultValue).slice().sort(),
```
Comment (scharf) [inline packages/moxb/src/routing/url-arg/UrlArgTypes.ts]: I do a `slice` here to not accidentally sort the `defaultValue`

**Comment by csillag** [comment]
> `ves` does not work with the master version of `moxb`, because it complains about `lodash.memoize`. This problem does probably not happen, if we would install `moxb` the normal way. But the combination of `lerna` and linked packages causes problems.
> 

That is a problem. I wonder what is the difference between `lodash.memoize` and any other npm dependencies? Anyway, if we can't solve the NPM problem, I can just as well copy the code directly to moxb. It's a single JS source file.

> However, I was looking why `moxb` needs `lodash.memoize`: it is needed by `UrlArgTypes.ts`.

Indeed. The reason we are using memoize is that we want to avoid unnecessary computations and react re-renders. Many of our reactivity components (mobx computations, react component properties) test for object identity, when triggering changes. So if every new reading of an URL argument returns a new object, this new object can trickle down all over the system, all the way to the UI, causing unwanted and useless calculations. We can stop all this by returning always the same array, as long as the input haven't changed.
 
> Looking at the tests I discovered some inconsistencies:
> 
> * sometimes the test requires the result of `parse()` to be the same, and sometimes it is not the same

Could you please help me find which parts are inconsistent?

> * for unordered arrays, the order should not matter when `parse()` is called (the result should always be the same)

This is not a real problem. Our only goal with memoize is that we want to avoid returning new array _constantly_, when the input data haven't changed. If the input data changes so that the order is now reversed, and we return a new array _once_, that will not cause a performance problem, since as long as the input data stays stable, we will keep returning the same (new) array.

It's true that we _could_ return the same array in these cases, but not doing so is only a missed opportunity for (marginal) performance optimization.

> * the `defaultValue` was ignored for attray types

That is only true for the parsers, but that's not a problem, since it is handled in another level of the code. (The actual UrlArg code.)

I reviewed the code and also tested with the browser, and I can confirm that when the URL argument is an empty string, the parsers are not called. The default value is returned instead, as designed. 

> So, I thought, I would fix it (see PR #47), but that turns out to be more complicated. Therefore, I decided to drop the constraint (and `lodash.memoize`), that the args are the same and reduced it to just be equal.

We _can_ do that, and it won't cause invalid results, but it might give us performance problems.
I have seen this happening in my previous project.

> I think there is a danger in `lodash.memoize` if we are not in a purely functional environment: if you change the returned array, then you change the cached value and all kind of evil things could happen:

Yes, that's true.
However, for all kinds of reasons, I think we _should_ do calculations in a functional manner.
(In most cases, anyway.)

So to solve this specific problem with memoize, I propose enforcing the immutability of the data, using one of the immutable libraries, like [seamless-immutable](https://github.com/rtfeldman/seamless-immutable).

**Comment by scharf** [comment]
> Indeed. The reason we are using memoize is that we want to avoid unnecessary computations and react re-renders. Many of our reactivity components (mobx computations, react component properties) test for object identity, when triggering changes. So if every new reading of an URL argument returns a new object, this new object can trickle down all over the system, all the way to the UI, causing unwanted and useless calculations. We can stop all this by returning always the same array, as long as the input haven't changed.

Where exactly does this happen? 

If we use [`@observable.struct`](https://mobx.js.org/refguide/modifiers.html) then a new assignment to a structurally equal value does not trigger a change. Caching is always problematic and I see no reason why we should add this complication if we can solve the potential problem with `mobx` built in mechanisms....

> Could you please help me find which parts are inconsistent?

When I wrote the comment, I did not fully understand why it has to be the same and therefore I expected  `URLARG_TYPE_UNORDERED_STRING_ARRAY` this test to succeed:

```ts
const v1 = parser('a,b,c', ['42']);
const v2 = parser('c,b,a', ['42']);
expect(v1).toBe(v2); /
```

> > the `defaultValue` was ignored for array types
> That is only true for the parsers, but that's not a problem, since it is handled in another level of the code. (The actual UrlArg code.)

Then we don't need the `defaultValue`  at all...

**Comment by scharf** [comment]
Changing [`BasicLocationManagerImpl._location`](https://github.com/moxb/moxb/blob/a9de2f563982e3635308b49015713c8d329a77df/packages/moxb/src/routing/location-manager/BasicLocationManagerImpl.ts#L63) to `@observable.struct` should solve the change problems, unless `_location` is manipulated in other ways than assigning new values.
